shell/common/shell.cc

```diff
@@ -2023,6 +2033,66 @@ bool Shell::OnServiceProtocolReloadAssetFonts(
   return true;
 }
 
+void Shell::AddView(int64_t view_id, const ViewportMetrics& viewport_metrics) {
+  TRACE_EVENT0("flutter", "Shell::AddView");
+  FML_DCHECK(is_set_up_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+  if (view_id == kFlutterImplicitViewId) {
+    FML_DLOG(ERROR) << "Unexpected request to add the implicit view #"
+                    << kFlutterImplicitViewId
+                    << ". This view should never be added. This call is no-op.";
+    return;
+  }
+
+  task_runners_.GetUITaskRunner()->PostTask([engine = engine_->GetWeakPtr(),  //
+                                             viewport_metrics,                //
+                                             view_id                          //
+  ] {
+    if (engine) {
+      engine->AddView(view_id, viewport_metrics);
+    }
+  });
+}
+
+void Shell::RemoveView(int64_t view_id) {
+  TRACE_EVENT0("flutter", "Shell::RemoveView");
+  FML_DCHECK(is_set_up_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+  if (view_id == kFlutterImplicitViewId) {
+    FML_DLOG(ERROR)
+        << "Unexpected request to remove the implicit view #"
+        << kFlutterImplicitViewId
+        << ". This view should never be removed. This call is no-op.";
+    return;
+  }
+
+  expected_frame_sizes_.erase(view_id);
+  fml::AutoResetWaitableEvent latch;
+  task_runners_.GetUITaskRunner()->PostTask(
+      [&latch,                                  //
+           & task_runners = task_runners_,      //
+       engine = engine_->GetWeakPtr(),          //
+       rasterizer = rasterizer_->GetWeakPtr(),  //
+       view_id                                  //
+  ] {
+        if (engine) {
+          engine->RemoveView(view_id);
+        }
+        // Don't wait for the raster task here, which only cleans up memory and
+        // does not affect functionality. Make sure it is done after Dart
+        // removes the view to avoid receiving another rasterization request
+        // that adds back the view record.
+        task_runners.GetRasterTaskRunner()->PostTask([rasterizer, view_id]() {
+          if (rasterizer) {
+            rasterizer->CollectView(view_id);
+          }
+        });
+        latch.Signal();
+      });
+
+  latch.Wait();
```

Comment:
Instead of blocking the platform thread, should this method accept a callback that is executed once the rasterizer has been updated?

Reply:
(Ah there's another comment I failed to address)
Yeah we might not need to wait for the UI task to finish. Let me remove the block first, and we'll see if we actually want a callback (we'll need a decision when we add the embedder API.)

Reply:
Actually I'm not sure if we need to block until this task to finish at all. Theoretically, if an embedder requests to remove a view, even if the view ID is referred in the future, the embedder can just silently ignore them. I think I'll remove the block for now, and we'll see if we need a callback in the future (we'll need to make a decision when we add the embedder API.)

Reply:
As a result I added the following doc:
```dart
  ///         This method returns immediately and does not wait for the task on
  ///         the UI thread to finish. This means that the Dart VM might still
  ///         send messages regarding this view ID for a short while, even
  ///         though this view ID is already invalid.
```